text/0000-command-ergonomics.md

```diff
@@ -0,0 +1,542 @@
+
+For example a process which exited nonzero
+probably printed to stderr;
+with `piped` we capture that, and represent both the stderr
+and the exit status as problems within the `SubprocessError`.
+
+```rust
+impl SubprocessError {
+    /// The program, if we know it.
+    fn program(&self) -> Option<&OsStr>;
+
+    /// The arguments, if we know them.
+    fn args(&self) -> Option<std::process::CommandArgs<'_>>;
+
+    /// If the stdout was captured in memory, the stdout data.
+    fn stdout_bytes(&self) -> Option<&[u8]>;
+
+    /// If the process exited and we collected its status, the exit status.
+    fn status(&self) -> Option<ExitStatus>;
+
+    /// If trouble included nonempty stderr, the captured stderr
+    fn stderr_bytes(&self) -> Option<&[u8]>;
+
+    /// If trouble included failure to spawn, the spawn error.
+    fn spawn_error() -> Option<&io::Error>;
+
+    /// If trouble included failure to talk to the child, the IO error.
+    ///
+    /// This might include problems which might be caused by child
+    /// misbehaviour.
+    fn communication_error() -> Option<&io::Error>;
+
+    /// If trouble included failed UTF-8 conversion.
+    fn utf8_error(&self) -> Option<&std::str::FromUtf8Error>;
+}
+```
+
+The `Display` implementation will print everything above
+including the command's arguments.
+The arguments will be escaped or quoted in some way that renders 
+a resulting error message unambiguous.
+
+### `impl From<SubprocessError> for io::Error`
+
+`SubprocessError` must be convertible to `io::Error`
+so that we can use it in `ChildOutputStream`'s
+`Read` implementation.
+This may also be convenient elsewhere.
+
+The `io::ErrorKind` for a `SubprocessError` will be:
+
+  * The `io::ErrorKind` from the spawn error, if any.
+
+  * Otherwise, a new kind `io::ErrorKind::ProcessFailed`,
+    which means that the subprocess itself failed.
+
+### Further necessary APIs for `SubprocessError`
+
+We also provide ways for this new error to be constructed,
+which will be needed by other lower level libraries besides std,
+notably async frameworks:
+
+```rust
+impl SubprocessError {
+    /// Makes a "blank" error which doesn't contain any useful information
+    ///
+    /// `has_problem()` will return `false` until one of the setters
+    /// is used to store an actual problem.
+    fn new_empty() -> Self { }
+
+    // If we keep ExitStatusError
+    fn from_exit_status_error(status: ExitStatusError) -> Self { }
+
+    fn set_program(&mut self, impl Into<OsString>);
+    fn set_args(&mut self, impl IntoIterator<Item=impl Into<OsString>>);
+    fn set_stdout_bytes(output: Option<Into<Box<[u8]>>>);
+
+    fn set_status(&mut self, status: ExitStatus);
+    fn set_stderr_bytes(&mut self, stderr: impl Into<Box<u8>>);
+    fn set_spawn_error(&mut self, error: Option<io::Error>);
+    fn set_communication_error(&mut self, error: Option<io::Error>);
+    fn set_utf8_error(&mut self, error: Option<std::str::FromUtf8Error>);
```

Comment:
I think too much of this is mutable -- either we should just make these be public fields (we have `get_` and `set_` methods for them...) or we should just allow constructing one of these from parts.

Additionally it's not clear to me how these interact. Up until now, I had imagined that there was an internal enumeration 

```rs
enum SubprocessErrorCause {
    Spawn(io::Error),
    Communication(io::Error),
    // Actually, I don't think we want this one, but anyway
    Utf8(str::Utf8Error),
    // ...
}
```
but this API has individual setters for each one, which implies it either is not that way, or they have perhaps-confusing interactions.

Does it make any sense for there to be more than one error here anway? Which would gets reported in `std::error::Error::source`?

Reply:
I have made this a transparent struct.  Making it opaque was indeed confusing.